client/src/components/Problem/ProblemShow.js

```diff
@@ -98,29 +108,57 @@ class ProblemShow extends Component {
         push(`/courses/${courseId}/assignments/${assignment.get('id')}`);
     }
 
+    onAccept = accepted => {
+        const { dispatch, problem } = this.props;
+
+        this.setState({
+            rejected: null
+        })
+
+        dispatch(submitSubmission(this.getProblemId(), accepted))
+    }
+
+    onReject = rejected => {
+        this.setState({
+            rejected
+        })
+    }
+
     render() {
         const test = this.props.test;
         const assignment = this.getAssignment();
         const problem = this.getProblem(assignment);
+        const { user } = this.props;
+        const { accepted, rejected } = this.state;
 
         return (
-                <div className="problemShow">
-                    <h1>
-                        <div className="backButtonAndText" onClick={ this.goBackToProblems }><i className="fa fa-angle-left backButton" aria-hidden="true"></i>{ assignment.get('name') } - { problem.get('name') }</div>
-                    </h1>
-                    {
+            <div className="problemShow">
+                <h1>
+                    <div className="backButtonAndText" onClick={ this.goBackToProblems }><i className="fa fa-angle-left backButton" aria-hidden="true"></i>{ assignment.get('name') } - { problem.get('name') }</div>
+                </h1>
+                {
                     test.getIn([problem.get('id'), 'fetched'])
-                    ? (
+                ? (
                     <TestList problemId={ problem.get('id') } />
-                    ) : null
-                    }
-                </div>
-                );
+                ) : null
+                }
+                {
+                    user.get('isAdmin')
+                ? (
+                    <div classname="submitSection">
+                        <label>Upload canonical solution:</label>
+                        <Filedrop onAccept={ this.onAccept } onReject={ this.onReject } accepted={ accepted } rejected={ rejected } accept=".java,.py" />
+                    </div>
+                ) : null
+                }
+            </div>
+        );
     }
 }
 
 export default connect(state => ({
     assignment: state.assignment,
     problem: state.problem,
-    test: state.test
+    test: state.test,
+    user: state.user,
```

Comment:
Seems all we care about is whether the user is admin. We can actually just pass in `isAdmin: state.user.get('isAdmin'),`. I'm trying to do this more React compares props to determine whether to re-render. Reducing the number/scope of props should theoretically mean fewer re-renders.

ALSO: please add prop validation for this at the top. It logs errors when required props are omitted, which really helps tracking down bugs.